### ISSUE TITLE
Feature request: sign release artifacts

### ISSUE THREAD

**Issue by mangin** [aws/aws-sam-cli#3751]
### Describe your idea/feature/enhancement
Dear team,
Could you please sign sha256sums of your artifacts and provide sha256sums as a separate sha256sums.asc file?

This little change will:
1. increases security of infrastructure of your customers
2. allows to check that an artifact release was built by an authorized user.

### Proposal

If you are interesting in it you can take a look on git-lfs team's implementation:
https://github.com/git-lfs/git-lfs/releases/tag/v3.1.2

This small bash script allows your customers to check that the archive is valid and was built by autorized user:

```sh
wget -O archive.tar "https://github.com/aws/aws-sam-cli/releases/download/v1.40.1/aws-sam-cli-linux-x86_64.zip" \
   && wget -O sha256sums.asc "https://github.com/aws/aws-sam-cli/releases/download/v1.40.1/sha256sums.asc" \
   && export GNUPGHOME="$(mktemp -d)" \
   && gpg --batch --verify sha256sums.asc \
   && export GITLFS_HASH_SUM=`cat sha256sums.asc | grep 'aws-sam-cli-linux-x86_64.zip' | cut -d ' ' -f1` \
   && echo "$GITLFS_HASH_SUM *aws-sam-cli-linux-x86_64.zip" | sha256sum --check --strict -
```

I hope that this issue are useful for your team!

**Comment by torresxb1** [comment]
Thanks for the feature request! We have a [roadmap](https://github.com/aws/aws-sam-cli/wiki/SAM-CLI-Roadmap) here, but we also try to take in popular feature requests based on 👍 from the community. We will be adding this to our intake for prioritization.

**Comment by torresxb1** [comment]
An update on this: our .`pkg` and `.msi` installers are already signed, but our linux artifacts are not. The team is having a discussion on this, but thumbs up on this issue would help us prioritize this.

**Comment by qingchm** [comment]
Thanks for the advice here! This request is targeting for a signing on the hash strings that we post on GH, which is also an area that we should definitely try to improve in a more secure and customer trusted direction. we will think about how to make the published hashes safer and provide updates here!

**Comment by qingchm** [comment]
@mangin Hey there after we considered your proposal here is what we want to provide: since for now our macOS installer and Windows installers are already signed, we will update our documentation to inform ways to verify those signatures besides the GitHub hashes. As for the Linux installer, we will sign with GnuPG and publish a public key. Each release we will also release a signature file that you can verify the integrity of our Linux ZIP using `gpg --verify aws-sam-cli-linux-x86_64.sig aws-sam-cli-linux-x86_64.zip` command. Let me know what you think about this solution!

**Comment by mangin** [comment]
@qingchm Hey. It should work.
As an alternative solution you can generate hash-sum file and a sign for it. This solution a little bit better scalable. Let's imagine that you decided to generate releases for different architecture.

In this case you should generate an additional file per architecture. For example:
aws-sam-cli-linux-x86_64.sig aws-sam-cli-linux-x86_64.zip
aws-sam-cli-linux-arm_64.sig aws-sam-cli-linux-arm_64.zip

With the alternative solution we wouldn't generate new file when we'll add new architecture.

**Comment by qingchm** [comment]
@mangin Yep I understand that storing all the hashes in one file is scalable in terms of number of files that we publish on GitHub, however we do think that signature itself is a more secure method than the string value hashes generated from sha256sum calls. We would really encourage customers to verify the signatures instead of relying on signed hashes for better manual hash sum check if that makes sense! For example for macOS installers we can run `codesign -dv —verbose=4 /usr/local/bin/sam` to verify the signature, which to me is a better way of checking the integrity than manually comparing the package hash to the published value. Hope that the explanation makes sense! And of course we can provide an additional signed hash file if it's a strong requirement from you, but overall I would encourage on directly verifying the signatures as signatures already use hash values to check for integrity!

**Comment by mangin** [comment]
@qingchm for me both variants are fine. I'm looking forward for any of them.

> signature itself is a more secure method than the string value hashes
I think it's not true. It's a really hard to modify your release + have the same hash sum. if it was easy we'll have a lot of problem in encryption algorithms.

But as I said for me it doesn't matter how you are going to sign the release. For me it's only matter to be able to check that the release engineer signed it.

**Comment by qingchm** [comment]
Thanks for the feedback, we will let you know when this new feature is available.

**Comment by hnnasit** [comment]
Hi @mangin, this feature has been shipped in SAM CLI [1.85.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.85.0). More details about verifying the integrity of Linux artifact can be found [here](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/reference-sam-cli-install-verify.html). Closing this issue, feel free to open a new issue if you have any questions.
